tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c

```diff
@@ -229,21 +243,608 @@ static void test_rx_on_key_refresh(void)
+		found = true;
+	}
+
+	if (found) {
+		k_sem_give(&observer_sem);
+	}
+}
+
+/* Listens to beacons for one beacon interval (10 seconds). */
+static bool wait_for_beacon(bool (*process_cb)(const uint8_t *net_id, void *ctx), void *ctx)
+{
+	struct bt_le_scan_param scan_param = {
+			.type       = BT_HCI_LE_SCAN_PASSIVE,
+			.options    = BT_LE_SCAN_OPT_NONE,
+			.interval   = BT_MESH_ADV_SCAN_UNIT(1000),
+			.window     = BT_MESH_ADV_SCAN_UNIT(1000)
+	};
+	bool received = false;
+	int err;
+
+	beacon.process_cb = process_cb;
+	beacon.user_ctx = ctx;
+
+	err = bt_le_scan_start(&scan_param, beacon_scan_cb);
+	if (err && err != -EALREADY) {
+		FAIL("starting scan failed (err %d)", err);
+	}
+
+	/* Listen to beacons ONLY for one beacon interval. */
+	err = k_sem_take(&observer_sem, BEACON_INTERVAL);
+	if (!err) {
+		received = true;
+	} else {
+		LOG_WRN("Didn't receive SNB in time (err: %d)", err);
+	}
+
+	err = bt_le_scan_stop();
+	if (err && err != -EALREADY) {
+		FAIL("stopping scan failed (err %d)", err);
+	}
+
+	/* Sleep a little to get to the next beacon interval. Otherwise, calling this function
+	 * again will catch the old beacon. This happens due to a known bug in legacy advertiser,
+	 * which transmits advertisements longer than should.
+	 */
+	k_sleep(K_SECONDS(1));
+
+	return received;
+}
+
+static void send_beacon(struct net_buf_simple *buf)
+{
+	struct bt_data ad;
+	int err;
+
+	ad.type = BT_DATA_MESH_BEACON;
+	ad.data = buf->data;
+	ad.data_len = buf->len;
+
+	err = bt_le_adv_start(BT_LE_ADV_NCONN, &ad, 1, NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)\n", err);
+	}
+
+	LOG_INF("Advertising started\n");
+
+	k_sleep(K_MSEC(100));
+
+	err = bt_le_adv_stop();
+	if (err) {
+		FAIL("Unable to stop advertising");
+	}
+}
+
+static void beacon_create(struct net_buf_simple *buf, const uint8_t *net_key, uint8_t flags,
+			  uint32_t iv_index)
+{
+	uint8_t beacon_key[16];
+	uint8_t net_id[8];
+	uint8_t auth[8];
+	int err;
+
+	err = bt_mesh_k3(net_key, net_id);
+	if (err) {
+		FAIL("Unable to generate Net ID");
+	}
+
+	err = bt_mesh_beacon_key(net_key, beacon_key);
+	if (err) {
+		FAIL("Unable to generate beacon key");
+	}
+
+	err = bt_mesh_beacon_auth(beacon_key, flags, net_id, iv_index, auth);
+	if (err) {
+		FAIL("Unable to generate auth value");
+	}
+
+	net_buf_simple_reset(buf);
+	net_buf_simple_add_u8(buf, 0x01 /* Secure Network Beacon */);
+	net_buf_simple_add_u8(buf, flags);
+	net_buf_simple_add_mem(buf, net_id, 8);
+	net_buf_simple_add_be32(buf, iv_index);
+	net_buf_simple_add_mem(buf, auth, 8);
+}
+
+/* Test reception of invalid beacons. */
+static void test_tx_invalid(void)
+{
+	NET_BUF_SIMPLE_DEFINE(buf, 22);
+	int err;
+
+	bt_mesh_test_cfg_set(&tx_cfg, 130);
+	k_sem_init(&observer_sem, 0, 1);
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Bluetooth init failed (err %d)", err);
+		return;
+	}
+
+	LOG_INF("Bluetooth initialized");
+
+	/* Let the node send the first beacon. */
```

Comment:
To clarify when you write `the node` you are reffering to the rx node? In either case it could be helpful to clarify the identity of the two participating nodes (tx/rx) more precicely.